Firestore/core/src/firebase/firestore/api/settings.h

```diff
@@ -0,0 +1,88 @@
+
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_SETTINGS_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_SETTINGS_H_
+
+#include <string>
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+/**
+ * Represents settings associated with a FirestoreClient.
+ *
+ * PORTING NOTE: We exclude the user callback std::executor in order to avoid
+ * ownership complexity.
+ */
+class Settings {
+ public:
+  Settings() = default;
+
+  void set_host(const std::string_view& value) {
```

Comment:
There is no `std::string_view` yet.

Since you're going to make copy when you retain the value, you can make this take a `std::string` and then move it. If the caller passes a temporary (or `std::move`s) this will save a copy.

```c++
void set_host(std::string value) {
  host_ = std::move(value);
}
```

Reply:
To add some background, `std::string_view` is from C++17. You could use `absl::string_view`, but since you're copying the value, Gil's suggestion is strictly better.

Reply:
Ahh, whoops!  So I guess we're stuck on C++11 but for some reason our compilers (whatever xcode, cmake, and travis are using) happily accept this, but we can't rely on it because we need to be compilable with other compilers that might not support it (like MSVC or something?)?  Is there no way to limit ourselves to c++11 features other than code review audits?

Anyway, I'll remove it but https://goto.google.com/totw/117 says:

> passing by const reference is simpler and safer, so it's still a good default choice. However, if you're working in an area that's known to be performance-sensitive, or your benchmarks show that you're spending too much time copying function parameters, passing by value can be a very useful tool to have in your toolbox.

This is very much not performance sensitive, so I'll just do `const std::string &`.  I'm actually not sure why I tried to use string_view in the first place.